src/concurrent_deque.hpp

```diff
@@ -0,0 +1,90 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <mutex>
+
+namespace realm {
+
+class ConcurrentDequeTimeout : public std::exception {
+  public:
+    ConcurrentDequeTimeout() : std::exception() {}
+};
+
+template <typename T>
+class ConcurrentDeque {
```

Comment:
Is this really necessary? Couldn't we just use dispatch queues and avoid the complexity?